cpu/gd32v/periph/i2c.c

```diff
@@ -0,0 +1,476 @@
+        }
+        DEBUG("[i2c] start: Address generated\n");
+        return ret;
+    }
+    return 0;
+}
+
+static int _is_sr1_mask_set(I2C_Type *i2c, uint32_t mask, uint8_t flags)
+{
+    DEBUG("[i2c] _is_sr1_mask_set: waiting to set %04X\n", (uint16_t)mask);
+    uint16_t tick = TICK_TIMEOUT;
+    while (tick--) {
+        uint32_t sr1 = i2c->STAT0;
+        if (sr1 & I2C_STAT0_AERR_Msk) {
+            DEBUG("[i2c] is_sr1_mask_set: NACK received\n");
+            i2c->STAT0 &= ~ERROR_FLAGS;
+            if (!(flags & I2C_NOSTOP)) {
+                _stop(i2c);
+            }
+            return -EIO;
+        }
+        if ((sr1 & I2C_STAT0_LOSTARB_Msk) || (sr1 & I2C_STAT0_BERR_Msk)) {
+            DEBUG("[i2c] is_sr1_mask_set: arb lost or bus ERROR_FLAGS\n");
+            i2c->STAT0 &= ~ERROR_FLAGS;
+            _stop(i2c);
+            return -EAGAIN;
+        }
+        if (sr1 & mask) {
+            i2c->STAT0 &= ~ERROR_FLAGS;
+            return 0;
+        }
+    }
+    /*
+    * If timeout occurs this means a problem that must be handled on a higher
+    * level.  A SWRST is recommended by the datasheet.
+    */
+    i2c->STAT0 &= ~ERROR_FLAGS;
+    _stop(i2c);
+    return -ETIMEDOUT;
+}
+
+static int _stop(I2C_Type *i2c)
+{
+    /* send STOP condition */
+    DEBUG("[i2c] stop: Generate stop condition\n");
+    i2c->CTL0 &= ~(I2C_CTL0_ACKEN_Msk);
+    i2c->CTL0 |= I2C_CTL0_STOP_Msk;
+    uint16_t tick = TICK_TIMEOUT;
+    while ((i2c->CTL0 & I2C_CTL0_STOP_Msk) && tick--) {}
+    if (!tick) {
+        return -ETIMEDOUT;
+    }
+    DEBUG("[i2c] stop: Stop condition succeeded\n");
+    if (_wait_for_bus(i2c) < 0) {
+        return -ETIMEDOUT;
+    }
+    DEBUG("[i2c] stop: Bus is free\n");
+    return 0;
+}
+
+static inline int _wait_for_bus(I2C_Type *i2c)
+{
+    uint16_t tick = TICK_TIMEOUT;
+    while ((i2c->STAT1 & I2C_STAT1_I2CBSY_Msk) && tick--) {}
+    if (!tick) {
+        return -ETIMEDOUT;
+    }
+    return 0;
+}
+
+static inline void _irq_handler(unsigned irq)
+{
+    i2c_t dev;
+
+    for (dev = 0; dev < I2C_NUMOF; dev++) {
+        if (i2c_config[dev].irqn == irq) {
+            break;
+        }
+    }
+
+    assert(dev < I2C_NUMOF);
+
+    I2C_Type *i2c = i2c_config[dev].dev;
+
+    assert(i2c != NULL);
+
+    unsigned state = i2c->STAT0;
+    DEBUG("\n\n### I2C ERROR OCCURRED ###\n");
+    DEBUG("status: %08x\n", state);
+    if (state & I2C_STAT0_OUERR_Msk) {
+        DEBUG("OVR\n");
+    }
+    if (state & I2C_STAT0_AERR_Msk) {
+        DEBUG("AF\n");
+    }
+    if (state & I2C_STAT0_LOSTARB_Msk) {
+        DEBUG("ARLO\n");
+    }
+    if (state & I2C_STAT0_BERR_Msk) {
+        DEBUG("BERR\n");
+    }
+    if (state & I2C_STAT0_PECERR_Msk) {
+        DEBUG("PECERR\n");
+    }
+    if (state & I2C_STAT0_SMBTO_Msk) {
+        DEBUG("TIMEOUT\n");
+    }
+    if (state & I2C_STAT0_SMBALT_Msk) {
+        DEBUG("SMBALERT\n");
+    }
+    core_panic(PANIC_GENERAL_ERROR, "I2C FAULT");
```

Comment:
I don't get this interrupt handler.
It's just printing the ISR flags, then panics?

Reply:
Yes, it really doesn't make sense. For example, the loss of arbitration is a normal situation. But this is the implementation of STM32F1. In an intermediate version I had already removed this `core_panic`, but I forgot to remove it in the final version.